Applications/Resnet/jni/cifar_dataloader.cpp

```diff
@@ -93,13 +118,65 @@ void RandomDataLoader::next(float **input, float **label, bool *last) {
 }
 
 Cifar100DataLoader::Cifar100DataLoader(const std::string &path, int batch_size,
-                                       int splits) {
-  /// NYI!
+                                       int splits) :
+  batch(batch_size),
+  current_iteration(0),
+  file(path, std::ios::binary | std::ios::ate) {
+  constexpr char error_msg[] = "failed to create dataloader, reason: ";
+
+  NNTR_THROW_IF(!file.good(), std::invalid_argument)
+    << error_msg << " Cannot open file";
+
+  auto pos = file.tellg();
+  NNTR_THROW_IF((pos % Cifar100DataLoader::SampleSize != 0),
+                std::invalid_argument)
+    << error_msg << " Given file does not align with the format";
+
+  auto data_size = pos / (Cifar100DataLoader::SampleSize * splits);
+  idxes = std::vector<unsigned int>(data_size);
+  std::cout << "path: " << path << '\n';
+  std::cout << "data_size: " << data_size << '\n';
+  std::iota(idxes.begin(), idxes.end(), 0);
+  std::shuffle(idxes.begin(), idxes.end(), rng);
+
+  iteration_per_epoch = data_size / batch_size;
 }
 
 void Cifar100DataLoader::next(float **input, float **label, bool *last) {
-  /// NYI!
-  *last = true;
+
+  auto fill_one_sample = [this](float *input_, float *label_, int index) {
+    NNTR_THROW_IF(!file.good(), std::invalid_argument)
+      << "file is not good, reason: " << std::strerror(errno);
+    file.seekg(index * Cifar100DataLoader::SampleSize + 1, std::ios_base::beg);
+
+    uint8_t current_label;
+    file.read(reinterpret_cast<char *>(&current_label), sizeof(uint8_t));
+    fillLabel(label_, Cifar100DataLoader::NumClass, current_label);
+
```

Comment:
do you assume that the dataset stores the data first or label first? it looks label data first. Could you check this? ( If you are using old dataset we used use, then the data is the first )

Reply:
This is directly using the raw data available from the official website. The format is slightly different, so this is using generator and instead of file based.